fuzzing/fuzzer_worker.go

```diff
@@ -301,105 +319,149 @@ func (fw *FuzzerWorker) testCallSequence() (calls.CallSequence, []ShrinkCallSequ
+	// Check if our verifier signalled that we met our conditions
+	validShrunkSequence := false
+	if len(possibleShrunkSequence) > 0 {
+		validShrunkSequence, err = shrinkRequest.VerifierFunction(fw, possibleShrunkSequence)
 		if err != nil {
-			return nil, err
+			return false, err
 		}
+	}
 
+	// After testing the sequence, we'll want to rollback changes to reset our testing state.
+	if err = fw.chain.RevertToBlockNumber(fw.testingBaseBlockNumber); err != nil {
+		return false, err
+	}
+	return validShrunkSequence, nil
+}
+
+// shrinkCallSequence takes a provided call sequence and attempts to shrink it by looking for redundant
+// calls which can be removed, and values which can be minimized, while continuing to satisfy the provided shrink
+// verifier.
+// Returns a call sequence that was optimized to include as little calls as possible to trigger the
+// expected conditions, or an error if one occurred.
+func (fw *FuzzerWorker) shrinkCallSequence(callSequence calls.CallSequence, shrinkRequest ShrinkCallSequenceRequest) (calls.CallSequence, error) {
+	// Define a variable to track our most optimized sequence across all optimization iterations.
+	optimizedSequence := callSequence
+
+	// First try to remove any calls we can. We go from start to end to avoid index shifting.
+	for i := 0; i < len(optimizedSequence); {
 		// If our fuzzer context is done, exit out immediately without results.
 		if utils.CheckContextDone(fw.fuzzer.ctx) {
 			return nil, nil
 		}
 
-		// Check if our verifier signalled that we met our conditions
-		validShrunkSequence := false
-		if len(testedPossibleShrunkSequence) > 0 {
-			validShrunkSequence, err = shrinkRequest.VerifierFunction(fw, testedPossibleShrunkSequence)
-			if err != nil {
-				return nil, err
-			}
+		// Recreate our current optimized sequence without the item at this index
+		possibleShrunkSequence, err := optimizedSequence.Clone()
+		if err != nil {
+			return nil, err
 		}
+		possibleShrunkSequence = append(possibleShrunkSequence[:i], possibleShrunkSequence[i+1:]...)
 
-		// After testing the sequence, we'll want to rollback changes to reset our testing state.
-		if err = fw.chain.RevertToBlockNumber(fw.testingBaseBlockNumber); err != nil {
+		// Test the shrunken sequence.
+		validShrunkSequence, err := fw.testShrunkenCallSequence(possibleShrunkSequence, shrinkRequest)
+		if err != nil {
 			return nil, err
 		}
 
 		// If this current sequence satisfied our conditions, set it as our optimized sequence.
 		if validShrunkSequence {
-			optimizedSequence = testedPossibleShrunkSequence
+			optimizedSequence = possibleShrunkSequence
 		} else {
 			// We didn't remove an item at this index, so we'll iterate to the next one.
 			i++
 		}
 	}
 
+	// Next try to shrink our values of every transaction a given number of rounds.
+	for i := 0; i < len(optimizedSequence); i++ {
+		for optimizationRound := 0; optimizationRound < 200; optimizationRound++ {
```

Comment:
I believe echidna does 5000 shrink rounds by default - is there a reason we are shooting for 200?

Reply:
Do we want to add a config option for it?

Reply:
What do you think? So basically, we try to optimize values for each call 200 times. So the work will be 200 * call sequence length.

I was thinking we should update this because the value shrinker uses probability to determine when to shrink, so we should always force it to shrink at least one value in the future (updating `abi_values.go`). So for now I was thinking this could be fixed because it's low and probably won't be a problem, but maybe it still would be good to add an option? I can quickly add it if you think so.

Reply:
I think if we are going to do another pass around the shrinking logic anyway, we can keep as is. i don't think it matters much either way. We can see the efficacy of 200 in practice and evaluate whether we want to increase, decrease, or allow the user the flexibility to choose.